builtInFunctions/baseAccountGuarder.go

```diff
@@ -0,0 +1,94 @@
+package builtInFunctions
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+)
+
+// BaseAccountGuarderArgs is a struct placeholder for
+// all necessary args to create a newBaseAccountGuarder
+type BaseAccountGuarderArgs struct {
+	GuardedAccountHandler vmcommon.GuardedAccountHandler
+	Marshaller            marshal.Marshalizer
+	EnableEpochsHandler   vmcommon.EnableEpochsHandler
+	FuncGasCost           uint64
+}
+
+type baseAccountGuarder struct {
+	baseActiveHandler
+	marshaller            marshal.Marshalizer
+	guardedAccountHandler vmcommon.GuardedAccountHandler
+
+	mutExecution sync.RWMutex
+	funcGasCost  uint64
+}
+
+func newBaseAccountGuarder(args BaseAccountGuarderArgs) (*baseAccountGuarder, error) {
+	if check.IfNil(args.Marshaller) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(args.EnableEpochsHandler) {
+		return nil, ErrNilEnableEpochsHandler
+	}
+	if check.IfNil(args.GuardedAccountHandler) {
+		return nil, ErrNilGuardedAccountHandler
+	}
+
+	accGuarder := &baseAccountGuarder{
+		funcGasCost:           args.FuncGasCost,
+		marshaller:            args.Marshaller,
+		mutExecution:          sync.RWMutex{},
+		guardedAccountHandler: args.GuardedAccountHandler,
+	}
+
+	accGuarder.activeHandler = args.EnableEpochsHandler.IsGuardAccountEnabled
+
+	return accGuarder, nil
+}
+
+func (baf *baseAccountGuarder) checkBaseAccountGuarderArgs(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+	expectedNoOfArgs uint32,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
```

Comment:
you should not care about the receiverAccount at all. the only check you need is vmInput.CallerAddr == vmInput.RecvAddress

Reply:
refactored & simplified the code here https://github.com/multiversx/mx-chain-vm-common-go/pull/193